compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -1474,11 +1474,14 @@ private void createRestFieldVarDefStmts(BLangTupleVariable parentTupleVariable,
             //     t[t.length()] = <T> tupleLiteral[$foreach$i];
             // }
             BLangExpression tupleExpr = parentTupleVariable.expr;
+            // Make sure the tupleExpr rest type can hold parent tuple's rest types
+            if (tupleExpr.type.tag == TypeTags.TUPLE) {
+                ((BTupleType) tupleExpr.type).restType = ((BArrayType) parentTupleVariable.restVariable.type).eType;
```

Comment:
We shouldn't be changing anything in the original tuple type, right? The following now panics at runtime.

```ballerina
public function main() {
    var [a, ...b] = tuple();
    [int, int, string...] m = <[int, int, string...]> tuple();
}

public function tuple() returns [int, int, string...] {
    return [1, 2, "hello", "world"];
}
```

```cmd
$ bal run test.bal 

Compiling source
        test.bal

Running executable

error: {ballerina}TypeCastError {"message":"incompatible types: '[int,int,int|string...]' cannot be cast to '[int,int,string...]'"}
        at test:main(test.bal:3)
```